src/ray/core_worker/reference_count.cc

```diff
@@ -1,28 +1,99 @@
   absl::MutexLock lock(&mutex_);
+  return AddBorrowedObjectInternal(object_id, outer_id, owner_id, owner_address);
+}
+
+bool ReferenceCounter::AddBorrowedObjectInternal(const ObjectID &object_id,
+                                                 const ObjectID &outer_id,
+                                                 const TaskID &owner_id,
+                                                 const rpc::Address &owner_address) {
   auto it = object_id_refs_.find(object_id);
   RAY_CHECK(it != object_id_refs_.end());
 
-  if (!it->second.owner.has_value()) {
-    it->second.owner = {owner_id, owner_address};
+  RAY_LOG(DEBUG) << "Adding borrowed object " << object_id;
+  // Skip adding this object as a borrower if we already have ownership info.
+  // If we already have ownership info, then either we are the owner or someone
+  // else already knows that we are a borrower.
+  if (it->second.owner.has_value()) {
+    RAY_LOG(DEBUG) << "Skipping add borrowed object " << object_id;
+    return false;
   }
+
+  it->second.owner = {owner_id, owner_address};
+
+  if (!outer_id.IsNil()) {
+    auto outer_it = object_id_refs_.find(outer_id);
+    if (outer_it == object_id_refs_.end()) {
+      outer_it = object_id_refs_.emplace(outer_id, Reference()).first;
+    }
+    if (outer_it->second.owned_by_us) {
+      RAY_LOG(DEBUG) << "Setting borrowed inner ID " << object_id
+                     << " contained_in_owned: " << outer_id;
+      it->second.contained_in_owned.insert(outer_id);
+    } else {
+      RAY_LOG(DEBUG) << "Setting borrowed inner ID " << object_id
+                     << " contained_in_borrowed: " << outer_id;
+      RAY_CHECK(!it->second.contained_in_borrowed_id.has_value());
+      it->second.contained_in_borrowed_id = outer_id;
+    }
+    outer_it->second.contains.insert(object_id);
+  }
+  return true;
 }
 
-void ReferenceCounter::AddOwnedObject(const ObjectID &object_id, const TaskID &owner_id,
+void ReferenceCounter::AddOwnedObject(const ObjectID &object_id,
+                                      const std::vector<ObjectID> &inner_ids,
+                                      const TaskID &owner_id,
                                       const rpc::Address &owner_address) {
+  RAY_LOG(DEBUG) << "Adding owned object " << object_id;
   absl::MutexLock lock(&mutex_);
   RAY_CHECK(object_id_refs_.count(object_id) == 0)
       << "Tried to create an owned object that already exists: " << object_id;
   // If the entry doesn't exist, we initialize the direct reference count to zero
   // because this corresponds to a submitted task whose return ObjectID will be created
   // in the frontend language, incrementing the reference count.
   object_id_refs_.emplace(object_id, Reference(owner_id, owner_address));
+  // Mark that this object ID contains other inner IDs. Then, we will not GC
+  // the inner objects until the outer object ID goes out of scope.
+  WrapObjectIdsInternal(object_id, inner_ids, owner_address, /*owned_by_us=*/true);
```

Comment:
The above comment is pretty clear. Maybe name this `MarkObjectIDsContained` or similar

Reply:
or `MarkObjectIDsWrapped`